From 57852f46fc18412a1bef3c21a6c250b3666a0e20 Mon Sep 17 00:00:00 2001
From: Andrei Cherechesu <andrei.cherechesu@nxp.com>
Date: Wed, 21 Apr 2021 21:53:19 +0300
Subject: [PATCH 3/4] imagebuilder: Add support for prepending path to file
 names

Added support for prepending path to file names in the final generated
u-boot script, for the use-case where we have the files in a separate
folder that can be accessed with a given $LOAD_CMD.

For example, we can have "fatload mmc 0:2" as LOAD_CMD but the
files would need to be loaded from the /boot folder within the 2nd
partition, not from the root ("/"). By specifying the "-p <path>"
parameter when running the script, paths like "/boot" can be
automatically prepended to the generated u-boot commands used
to load the files in board's memory.

Additionally skipped symbolic links resolving, since the file
names that we pass via configs are symlinks, but we want the
tool to generate a script with the exact same names that we
pass to it.

Signed-off-by: Andrei Cherechesu <andrei.cherechesu@nxp.com>
---
 scripts/uboot-script-gen | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/scripts/uboot-script-gen b/scripts/uboot-script-gen
index 40dc69d..2ca481d 100755
--- a/scripts/uboot-script-gen
+++ b/scripts/uboot-script-gen
@@ -125,9 +125,13 @@ function load_file()
 {
     local filename=$1
 
-    local relative_path="$(realpath --relative-to=$PWD $filename)"
+    local relative_path="$(realpath --no-symlinks --relative-to=$PWD $filename)"
 
-    echo "$LOAD_CMD $memaddr $relative_path" >> $UBOOT_SOURCE
+    if [ -z "$PREPEND_PATH" ]; then
+        echo "$LOAD_CMD $memaddr $relative_path" >> $UBOOT_SOURCE
+    else
+        echo "$LOAD_CMD $memaddr $PREPEND_PATH/$relative_path" >> $UBOOT_SOURCE
+    fi
     add_size $filename
 }
 
@@ -200,7 +204,7 @@ function print_help
 {
     script=`basename "$0"`
     echo "usage:"
-    echo "	$script -c CONFIG_FILE -t UBOOT_TYPE -d DIRECTORY [-o FILE]"
+    echo "	$script -c CONFIG_FILE -t UBOOT_TYPE -d DIRECTORY [-o FILE] [-p PREPEND_PATH]"
     echo "	$script -h"
     echo "where:"
     echo "	CONFIG_FILE - configuration file"
@@ -211,6 +215,7 @@ function print_help
     echo "		< > - used for uboot load commands"
     echo "	DIRECTORY - root directory where the files of CONFIG_FILE are located"
     echo "	FILE - output filename for the uboot script and its source, overrides option in CONFIG_FILE"
+    echo "	PREPEND_PATH - path to be appended before file names to match deploy location within rootfs"
     echo "	-h - prints out the help message and exits "
     echo "Defaults:"
     echo "	CONFIG_FILE=$cfg_file, UBOOT_TYPE=\"LOAD_CMD\" env var, DIRECTORY=$uboot_dir"
@@ -218,7 +223,7 @@ function print_help
     echo "	$script -c ../config -d ./build42 -t \"scsi load 1:1\""
 }
 
-while getopts ":c:t:d:ho:" opt; do
+while getopts ":c:t:d:ho:p:" opt; do
     case ${opt} in
     t )
         case $OPTARG in
@@ -245,6 +250,9 @@ while getopts ":c:t:d:ho:" opt; do
     o )
         UBOOT_SCRIPT_ARG=$OPTARG
         ;;
+    p )
+        PREPEND_PATH="$OPTARG"
+        ;;
     h )
         print_help
         exit 0
-- 
2.7.4

